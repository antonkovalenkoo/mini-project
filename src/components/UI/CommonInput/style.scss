.commonInput {
  position: relative;

  &__input {
    width: 278px;
    height: 30px;
    font-size: 15px;
    padding: 0 10px;
    border-radius: 8px;
    background-color: $input-color-primary;
    border: 1px solid $input-color-secondary;
  
    &:focus {
      outline: 1px solid $input-color-secondary;
    }
  }
  
  &__icon {
    position: absolute;
    height: 23px;
    width: 23px;
    padding: 5px;
  }
}

.padding-big {
  width: 230px;
  padding: 0 34px;
}

.postfix {
  transform: translateX(-100%);
}
