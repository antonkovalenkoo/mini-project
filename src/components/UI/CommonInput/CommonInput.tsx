import { FC } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ICommonInputProps } from '../../../models/index';
import './style.scss';

const CommonInput: FC<ICommonInputProps> = ({ 
  prefixIcon,
  postfixIcon,
  className,
  style,
  ...props
}) => {
  return (
    <div className='commonInput'>
      {prefixIcon && 
        <FontAwesomeIcon
          title='prefix'
          className='commonInput__icon' 
          icon={prefixIcon} 
        />
      }
      <input 
        {...props}
        className= {`commonInput__input ${prefixIcon && 'padding-big'}`}
      />
      {postfixIcon && 
        <FontAwesomeIcon
          title='postfix'
          className='commonInput__icon postfix' 
          icon={postfixIcon} 
        />
      }
    </div>
  )
}

export default CommonInput;