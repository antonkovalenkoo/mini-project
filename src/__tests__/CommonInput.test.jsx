import '@testing-library/jest-dom';
import { render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import CommonInput from "../components/UI/CommonInput/CommonInput";
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

const props = {
  type: 'text',
  placeholder: "Say Hello!",
  value: 'Hello!',
  prefixIcon: faCoffee,
  postfixIcon: faCoffee 
}

describe('CommonInput component', () => {
  test('CommonInput renders without props', () => {
    render(<CommonInput />);
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  test('CommonInput renders with prefix icon', () => {
    render(
      <CommonInput prefixIcon={faCoffee} />
    );;

    expect(screen.getByTitle('prefix')).toBeInTheDocument();
  });

  test('CommonInput renders with postfix icon', () => {
    render(
      <CommonInput postfixIcon={faCoffee} />
    );;

    expect(screen.getByTitle('postfix')).toBeInTheDocument();
  });

  test('CommonInput renders with both prefix and postfix icons', () => {
    render(
      <CommonInput
        postfixIcon={faCoffee}
        prefixIcon={faCoffee}
      />
    );;

    expect(screen.getByTitle('prefix')).toBeInTheDocument();
    expect(screen.getByTitle('postfix')).toBeInTheDocument();
  });

  test('CommonInput onChange passed to input', () => {
    const onChangeHandler = jest.fn();

    render(
      <CommonInput
        {...props}
        onChange={onChangeHandler}
      />
    );

    userEvent.type(screen.getByRole('textbox'), 'testText');
    expect(onChangeHandler).toHaveBeenCalledTimes(8)
  });

  test('CommonInput pass all input props to input', () => {
    render(
      <CommonInput {...props} />
    );

    const inputElement = screen.getByRole('textbox');
    expect(inputElement).toMatchSnapshot();

    const passingProps = Object.entries(props).filter(
      ([key, value]) => !['prefixIcon', 'postfixIcon'].includes(key)
    );

    passingProps.forEach(([key, value]) => {
      expect(inputElement.getAttribute(key)).toEqual(value)
    });
  });
})
