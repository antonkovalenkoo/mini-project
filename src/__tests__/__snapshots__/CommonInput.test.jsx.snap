// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonInput component CommonInput pass all input props to input 1`] = `
<input
  class="commonInput__input padding-big"
  placeholder="Say Hello!"
  type="text"
  value="Hello!"
/>
`;
